File: RailgunVB/Core/Managers/MusicManager.vb
Imports System.Text
Imports AudioChord
Imports Discord
Imports Microsoft.Extensions.DependencyInjection
Imports RailgunVB.Core.Logging
Imports RailgunVB.Core.Utilities
Imports TreeDiagram
Imports TreeDiagram.Models.Server

Namespace Core.Managers
    
    Public Class MusicManager
        
        Private ReadOnly _log As Log
        Private ReadOnly _commandUtils As CommandUtils
        Private ReadOnly _services As IServiceProvider
        Private ReadOnly _musicService As MusicService
        
        Public Sub New(services As IServiceProvider)
            _log = services.GetService(Of Log)
            _commandUtils = services.GetService(Of CommandUtils)
            _musicService = services.GetService(Of MusicService)
            _services = services
        End Sub
        
        Public Async Function AddYoutubeSongsAsync(urlList As List(Of String), 
                                                   tc As ITextChannel) As Task
            Dim playlist As Playlist
            
            Using scope As IServiceScope = _services.CreateScope()
                Dim context As TreeDiagramContext = scope.ServiceProvider.GetService(Of TreeDiagramContext)
                Dim data As ServerMusic = Await context.ServerMusics.GetOrCreateAsync(tc.GuildId)
                playlist = Await _commandUtils.GetPlaylistAsync(data)
            End Using
            
            Dim response As IUserMessage
            Dim playlistModified = False
                        
            For Each url As String In urlList
                response = Await tc.SendMessageAsync($"{Format.Bold("Processing :")} <{url}>...")
                    
                Dim cleanUrl As String = url.Trim(" "c, "<"c, ">"c)
                Dim videoId As String = String.Empty
                Dim song As ISong = Nothing
                            
                If Not (_musicService.Youtube.TryParseYoutubeUrl(url, videoId))
                    Await response.ModifyAsync(
                        Sub(properties) properties.Content = $"{Format.Bold("Invalid Url :")} <{cleanUrl}>")
                    Continue For
                ElseIf Await _musicService.TryGetSongAsync(New SongId("YOUTUBE", videoId), Sub(out) song = out)
                    If playlist.Songs.Contains(song.Id)
                        Await response.ModifyAsync(Sub(properties) properties.Content = 
                            $"{Format.Bold("Already Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}")
                    Else 
                        playlist.Songs.Add(song.Id)
                        playlistModified = True
                        Await response.ModifyAsync(Sub(properties) properties.Content = 
                            $"{Format.Bold("Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}")
                    End If
                                
                    Continue For
                End If
                            
                Try
                    song = Await _musicService.Youtube.DownloadAsync(New Uri(url))
                    playlist.Songs.Add(song.Id)
                    playlistModified = True
                    Await response.ModifyAsync(Sub(properties) properties.Content = 
                        $"{Format.Bold("Encoded & Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}")
                Catch ex As Exception
                    response.ModifyAsync(Sub(properties) properties.Content = 
                        $"{Format.Bold("Failed To Install :")} (<{cleanUrl}>), {ex.Message}").GetAwaiter()
                End Try
            Next
                
            If playlistModified Then Await _musicService.Playlist.UpdateAsync(playlist)
                
            Await tc.SendMessageAsync("Done!")
        End Function
        
        Public Async Function ProcessYoutubePlaylistAsync(playlist As Playlist, resolvingPlaylist As ResolvingPlaylist, 
                                                          tc As ITextChannel) As Task
            Dim alreadyInstalled = 0
            Dim failed = 0
            
            For Each songTask In resolvingPlaylist.Songs
                Try
                    Dim song As ISong = songTask.Result
                            
                    If playlist.Songs.Contains(song.Id)
                        alreadyInstalled += 1
                        Continue For
                    End If
                    
                    playlist.Songs.Add(song.Id)
                    
                    Await _musicService.Playlist.UpdateAsync(playlist)
                Catch
                    failed += 1
                End Try
            Next
            
            Dim output As New StringBuilder
            
            output.AppendLine("Finished Processing YouTube Playlist! Results...") _
                .AppendFormat(
                "Already Installed : {0} || Imported From Repository : {1} || Newly Encoded : {2} || Failed : {3}",
                    Format.Bold(alreadyInstalled),
                    Format.Bold(resolvingPlaylist.ExistingSongs),
                    Format.Bold(resolvingPlaylist.Songs.Count - resolvingPlaylist.ExistingSongs - failed),
                    Format.Bold(failed)
                )
            
            Await tc.SendMessageAsync(output.ToString())
        End Function
        
        Public Async Function YoutubePlaylistStatusUpdatedAsync(tc As ITextChannel, status As SongProcessStatus, 
                                                                data As ServerMusic) As Task
            Select status.Status
                Case SongStatus.Errored
                    Dim output As SongProcessError = status
                    Dim url As String = "https://youtu.be/" + output.Id.SourceId
                    
                    Await tc.SendMessageAsync(
                        $"{Format.Bold("Failed To Install :")} (<{url}>), {output.Exceptions.Message}")
                    
                    Dim logOutput As New StringBuilder
                    
                    logOutput.AppendFormat("<{0} ({1})> Process Failure!", tc.Guild.Name, tc.GuildId).AppendLine() _ 
                        .AppendFormat("{0} - {1}", url, output.Exceptions.Message)
                    
                    Await _log.LogToBotLogAsync(logOutput.ToString(), BotLogType.AudioChord)
                    Exit Select
                Case SongStatus.Processed
                    Dim output As SongProcessResult = status
                    Dim song As ISong = Await output.Result
                    Dim playlist As Playlist = Await _commandUtils.GetPlaylistAsync(data)
                    
                    playlist.Songs.Add(song.Id)
                    Await _musicService.Playlist.UpdateAsync(playlist)
                    Await tc.SendMessageAsync(
                        $"{Format.Bold("Encoded & Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}")
                    
                    Dim logOutput As New StringBuilder
                    
                    logOutput.AppendFormat("<{0} ({1})> Processed Song!", tc.Guild.Name, tc.GuildId).AppendLine() _ 
                        .AppendFormat("{0} <{1}> - {2}", song.Id.ToString(), song.Metadata.Length.ToString(), 
                                      song.Metadata.Name)
                    
                    Await _log.LogToBotLogAsync(logOutput.ToString(), BotLogType.AudioChord)
                    Exit Select
            End Select
            
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Core/Managers/MusicManager.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using AudioChord;
using Discord;
using Microsoft.Extensions.DependencyInjection;
using Railgun.Core.Logging;
using Railgun.Core.Utilities;
using TreeDiagram;
using TreeDiagram.Models.Server;

namespace Railgun.Core.Managers
{
    public class MusicManager
    {
        private readonly IServiceProvider _services;
        private readonly Log _log;
        private readonly CommandUtils _commandUtils;
        private readonly MusicService _musicService;

        public MusicManager(IServiceProvider services) {
            _services = services;

            _log = _services.GetService<Log>();
            _commandUtils = _services.GetService<CommandUtils>();
            _musicService = _services.GetService<MusicService>();
        }

        public async Task AddYoutubeSongsAsync(List<string> urls, ITextChannel tc) {
            Playlist playlist;

            using (var scope = _services.CreateScope()) {
                var db = scope.ServiceProvider.GetService<TreeDiagramContext>();
                var data = await db.ServerMusics.GetOrCreateAsync(tc.GuildId);
                playlist = await _commandUtils.GetPlaylistAsync(data);

                await db.SaveChangesAsync();
            }

            var playlistModified = false;

            foreach (var url in urls) {
                var response = await tc.SendMessageAsync($"{Format.Bold("Processing :")} <{url}>...");
                var cleanUrl = url.Trim(' ', '<', '>');
                var videoId = string.Empty;
                ISong song = null;

                if (!_musicService.Youtube.TryParseYoutubeUrl(url, out videoId)) {
                    await response.ModifyAsync(properties => properties.Content = $"{Format.Bold("Invalid Url :")} <{cleanUrl}>");
                    continue;
                } else if (await _musicService.TryGetSongAsync(new SongId("YOUTUBE", videoId), result => song = result)) {
                    if (playlist.Songs.Contains(song.Id))
                        await response.ModifyAsync(properties => properties.Content = $"{Format.Bold("Already Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}");
                    else {
                        playlist.Songs.Add(song.Id);
                        playlistModified = true;

                        await response.ModifyAsync(properties => properties.Content = $"{Format.Bold("Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}");
                    }

                    continue;
                }

                try {
                    song = await _musicService.Youtube.DownloadAsync(new Uri(url));
                    playlist.Songs.Add(song.Id);
                    playlistModified = true;

                    await response.ModifyAsync(properties => properties.Content = $"{Format.Bold("Encoded & Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}");
                } catch (Exception ex) {
                    await response.ModifyAsync(properties => properties.Content = $"{Format.Bold("Failed To Install :")} (<{cleanUrl}>), {ex.Message}");
                }
            }

            if (playlistModified) await _musicService.Playlist.UpdateAsync(playlist);

            await tc.SendMessageAsync("Done!");
        }

        public async Task ProcessYoutubePlaylistAsync(Playlist playlist, ResolvingPlaylist resolvingPlaylist, ITextChannel tc) {
            var alreadyInstalled = 0;
            var failed = 0;

            foreach (var songTask in resolvingPlaylist.Songs) {
                try {
                    var song = songTask.Result;

                    if (playlist.Songs.Contains(song.Id)) {
                        alreadyInstalled++;
                        continue;
                    }

                    playlist.Songs.Add(song.Id);

                    await _musicService.Playlist.UpdateAsync(playlist);
                } catch { failed++; }
            }

            var output = new StringBuilder()
                .AppendLine("Finished Processing YouTube Playlist! Results...")
                .AppendFormat(
                "Already Installed : {0} || Imported From Repository : {1} || Newly Encoded : {2} || Failed : {3}",
                    Format.Bold(alreadyInstalled.ToString()),
                    Format.Bold(resolvingPlaylist.ExistingSongs.ToString()),
                    Format.Bold(((resolvingPlaylist.Songs.Count - resolvingPlaylist.ExistingSongs) - failed).ToString()),
                    Format.Bold(failed.ToString())
                );
            
            await tc.SendMessageAsync(output.ToString());
        }

        public async Task YoutubePlaylistStatusUpdatedAsync(ITextChannel tc, SongProcessStatus status, ServerMusic data) {
            switch (status.Status) {
                case SongStatus.Errored:
                    {
                        var output = (SongProcessError)status;
                        var url = "https://youtu.be/" + output.Id.SourceId;
                        var logOutput = new StringBuilder()
                                .AppendFormat("<{0} <{1}>> Process Failure!", tc.Guild.Name, tc.GuildId).AppendLine()
                                .AppendFormat("{0} - {1}", url, output.Exceptions.Message);

                        try {
                            await tc.SendMessageAsync($"{Format.Bold("Failed To Install :")} (<{url}>), {output.Exceptions.Message}");
                        } catch (ArgumentException ex) {
                            await _log.LogToConsoleAsync(new LogMessage(LogSeverity.Warning, "Music Manager", "Missing Playlist", ex));
                        } catch (Exception ex) {
                            await _log.LogToConsoleAsync(new LogMessage(LogSeverity.Warning, "Music Manager", "Missing TC", ex));
                        }

                        await _log.LogToBotLogAsync(logOutput.ToString(), BotLogType.AudioChord);
                    }
                    break;
                case SongStatus.Processed:
                    {
                        var output = (SongProcessResult)status;
                        var song = await output.Result;
                        
                        var logOutput = new StringBuilder()
                            .AppendFormat("<{0} ({1})> Processed Song!", tc.Guild.Name, tc.GuildId).AppendLine()
                            .AppendFormat("{0} <{1}> - {2}", song.Id.ToString(), song.Metadata.Length.ToString(), song.Metadata.Name);

                        try {
                            var playlist = await _commandUtils.GetPlaylistAsync(data);

                            playlist.Songs.Add(song.Id);

                            await _musicService.Playlist.UpdateAsync(playlist);
                            await tc.SendMessageAsync($"{Format.Bold("Encoded & Installed :")} ({song.Id.ToString()}) {song.Metadata.Name}");
                        } catch (ArgumentException ex) {
                            await _log.LogToConsoleAsync(new LogMessage(LogSeverity.Warning, "Music Manager", "Missing Playlist", ex));
                        } catch (Exception ex) {
                            await _log.LogToConsoleAsync(new LogMessage(LogSeverity.Warning, "Music Manager", "Missing TC", ex));
                        }

                        await _log.LogToBotLogAsync(logOutput.ToString(), BotLogType.AudioChord);
                    }
                    break;
            }
        }
    }
}